fe/fe-core/src/main/java/com/starrocks/qe/scheduler/dag/PhasedExecutionSchedule.java

```diff
@@ -0,0 +1,449 @@
+                final int groupId = mergedUnionFind.getGroupIdOrAdd(childFragmentId);
+                Set<ExecutionFragment> groupFragments = groups.computeIfAbsent(groupId, k -> Sets.newLinkedHashSet());
+                int degree =
+                        scheduleFragmentInDegrees.compute(childFragmentId, (k, v) -> Preconditions.checkNotNull(v) - 1);
+                // we won't schedule the fragment that in-degree neq 0
+                if (degree == 0) {
+                    groupFragments.add(child);
+                } else {
+                    waitScheduleFragments.add(child.getFragmentId());
+                }
+            }
+
+            if (groups.size() != fragment.childrenSize()) {
+                List<PackedExecutionFragment> fragments =  Lists.newArrayList();
+                for (int i = fragment.childrenSize() - 1; i >= 0; i--) {
+                    final ExecutionFragment child = sequenceMap.get(fragment.getFragmentId()).getAt(i);
+                    final PlanFragmentId childFragmentId = child.getFragmentId();
+                    final int groupId = mergedUnionFind.getGroupIdOrAdd(childFragmentId);
+                    final Set<ExecutionFragment> groupFragments = groups.get(groupId);
+                    // The current fragment is in the same group as the prev fragment and
+                    // does not need to be processed.
+                    if (processedFragments.contains(childFragmentId)) {
+                        continue;
+                    }
+
+                    fragments.add(new PackedExecutionFragment(groupFragments));
+
+                    for (ExecutionFragment groupFragment : groupFragments) {
+                        processedFragments.add(groupFragment.getFragmentId());
+                    }
+
+                    // child that index > 0 needs to schedule the next round of fragments.
+                    // eg: each node is a fragment
+                    //           Union
+                    //            │
+                    //            │
+                    //            │
+                    //  ┌─────────┼─────────┐
+                    //  │         │         │
+                    //  ▼         ▼         ▼
+                    // AGG(1)    AGG(2)    AGG(3)
+                    //  │         │         │
+                    //  │         │         │
+                    //  ▼         ▼         ▼
+                    //SCAN(1)   SCAN(2)   SCAN(3)
+                    //
+                    // in this case when AGG2/AGG3 fragment finished we should trigger next schedule
+                    if (i != 0) {
+                        child.setNeedReportFragmentFinish(true);
+                    }
+                }
+                for (int i = fragments.size() - 1; i >= 0; i--) {
+                    stack.push(fragments.get(i));
+                }
+            } else {
+                for (int i = 0; i < fragment.childrenSize(); i++) {
```

Comment:
Why the traverse orders between L356 and L360 are different?

Reply:
Here we reverse the order because we need to schedule the rightmost child first.
fragments = [3,2,1]
```
for (int i = fragments.size() - 1; i >= 0; i--) {
    stack.push(fragments.get(i));
}
```
then add it to stack
stack = [1,2,3]

then we could schedule 3 firstly

Reply:
in L360. 

We'll press the stacks in turn, and they'll become stack=[1,2,3], and here we're also scheduling 3 first.